source/Octopus.Tentacle.Tests.Integration/WorkspaceCleanerTests.cs

```diff
@@ -0,0 +1,183 @@
+                })
+                .WithTentacleServiceDecorator(new TentacleServiceDecoratorBuilder()
+                    .DecorateScriptServiceV2With(d =>
+                        d.BeforeCompleteScript(async (_, _) =>
+                            {
+                                Directory.Exists(workspaceDirectory).Should().BeTrue("Workspace should exist before we complete");
+
+                                await Task.Delay(2 * 500, CancellationToken);
+
+                                Directory.Exists(workspaceDirectory).Should().BeTrue("Workspace should not have been cleaned up");
+                            })
+                            .Build())
+                    .Build())
+                .Build(CancellationToken);
+
+            workspaceDirectory = GetWorkspaceDirectoryPath(clientAndTentacle.RunningTentacle.HomeDirectory, startScriptCommand.ScriptTicket.TaskId);
+
+            await clientAndTentacle.TentacleClient.ExecuteScript(startScriptCommand, CancellationToken, null, new InMemoryLog());
+            
+            Directory.Exists(workspaceDirectory).Should().BeFalse("Workspace should be naturally cleaned up after completion");
+        }
+        
+        [Test]
+        [TentacleConfigurations(testDefaultTentacleRuntimeOnly: true)]
+        public async Task WhenTentacleClientCrashesAndTentacleNeverGotToldToComplete_ThenWorkspaceIsConsideredRunning_AndWillNotBeDeleted(TentacleConfigurationTestCase tentacleConfigurationTestCase)
+        {
+            var cleanerDelay = TimeSpan.FromMilliseconds(500);
+            var deleteWorkspacesOlderThan = TimeSpan.FromMilliseconds(500);
+
+            var startScriptCommand = new StartScriptCommandV2Builder().WithScriptBody(b => b.Print("Hello")).Build();
+
+            await using var clientAndTentacle = await tentacleConfigurationTestCase.CreateBuilder()
+                .WithTentacle(b =>
+                {
+                    b.WithRunTentacleEnvironmentVariable(WorkspaceCleanerConfiguration.CleanerDelayEnvironmentVariableName, cleanerDelay)
+                        .WithRunTentacleEnvironmentVariable(WorkspaceCleanerConfiguration.DeleteWorkspacesOlderThanTimeSpanEnvironmentVariableName, deleteWorkspacesOlderThan);
+                })
+                .WithTentacleServiceDecorator(new TentacleServiceDecoratorBuilder()
+                    .DecorateScriptServiceV2With(d =>
+                        d.BeforeCompleteScript((_, _) => throw new NotImplementedException("Force failure to simulate tentacle client crashing, and ensure we do not completion"))
+                        .Build())
+                    .Build())
+                .Build(CancellationToken);
+            
+            await AssertionExtensions
+                .Should(() => clientAndTentacle.TentacleClient.ExecuteScript(startScriptCommand, CancellationToken, null, new InMemoryLog()))
+                .ThrowAsync<NotImplementedException>();
+
+            await Task.Delay(1000, CancellationToken);
```

Comment:
nit: more deterministic way of ensuing the cleaner has run

Reply:
Have done the same thing as above for now, until we determine if we want to do more (e.g. log querying solution)